### PR TITLE
Feature/local stack tests

### PR DESCRIPTION
**Pull request by PierreJeanjacquot**
# Setup a local stack for tests

- [x] local test stack
  - [x] healthchecks
  - [x] start/stop script for the stack
- [x] CI test stack
  - [x] healthchecks
- [x] use local stack in tests

## Stack:

- blockchain
  - `bellecour-fork`
- core stack
  - `market-api` & `market-watcher`
  - `sms`
  - `result-proxy`
- dataprotector components 
  - `graphnode` with Dataprotector subgraph
  - `ipfs`
- ...and services deps (mutualized when possible)

The stack is configured by the `prepare-test-env.js` script to get the current block of the bellecour node to fork from.

### Local:

Before running tests locally, you must configure and start the local test stack by running

```sh
npm run start-test-stack
```
You will get a local stack running in docker.

Run your tests as usual
```sh
npm run tests
```

When finished you can teardown the local test stack
```sh
npm run stop-test-stack
```

If you need to refresh the stack (basically if the bellecour-fork restart, it will lose the history and components will be desynced), run the `start-test-stack` once again, this will teardown any existing stack and recreate a new one based on the current blockchain state.

### CI:

Do the same things as local but in a way we can do it in drone.

### CODE REVIEW

File: packages/sdk/.drone.yml [hunk 1]
```diff
@@ -7,6 +7,216 @@ trigger:
...
+      MONGO_HOST: result-proxy-mongo
+      MONGO_PORT: 13202
+      IEXEC_IPFS_HOST: ipfs
+    expose:
+      - 13200
+    commands:
+      # bellecour-fork healthcheck
+      # result-proxy-mongo healthcheck
+      # ipfs healthcheck
+      - sleep 30
```
Comment (PierreJeanjacquot) [inline packages/sdk/.drone.yml:134]: no access to `nc` or `curl` in this image, so the classic open port check is replaced by `sleep 30`

File: packages/sdk/.drone.yml [hunk 2]
```diff
@@ -7,6 +7,216 @@ trigger:
...
+      IEXEC_IGNORED_SGX_ADVISORIES: INTEL-SA-00161,INTEL-SA-00289,INTEL-SA-00334,INTEL-SA-00381,INTEL-SA-00389,INTEL-SA-00220,INTEL-SA-00270,INTEL-SA-00293,INTEL-SA-00320,INTEL-SA-00329,INTEL-SA-00477
+      IEXEC_SCONE_TOLERATED_INSECURE_OPTIONS: debug-mode,hyperthreading,outdated-tcb
+      IEXEC_SMS_DISPLAY_DEBUG_SESSION: 'true'
+      IEXEC_SCONE_CAS_HOST: foo
+      IEXEC_SMS_IMAGE_LAS_IMAGE: foo
+    expose:
+      - 13300
+    commands:
+      # bellecour-fork healthcheck
+      - sleep 30
```
Comment (PierreJeanjacquot) [inline packages/sdk/.drone.yml:114]: no access to `nc` or `curl` in this image, so the classic open port check is replaced by `sleep 30`

File: packages/sdk/package.json [hunk 3]
```diff
@@ -19,7 +19,9 @@
     "lint": "eslint .",
     "lint:fix": "eslint . --fix",
     "format": "prettier --write \"(src|tests)/**/*.ts\"",
-    "check-format": "prettier --check \"(src|tests)/**/*.ts|tests/**/*.ts\""
+    "check-format": "prettier --check \"(src|tests)/**/*.ts|tests/**/*.ts\"",
+    "stop-test-stack": "cd tests && docker compose down --remove-orphans",
+    "start-test-stack": "cd tests && npm run stop-test-stack && node prepare-test-env.js && docker compose up -d"
```
Comment (PierreJeanjacquot) [inline packages/sdk/package.json:24]: start and stop scripts (require docker compose)

File: packages/sdk/src/dataProtector/processProtectedData.ts [hunk 4]
```diff
@@ -24,14 +25,15 @@ export const processProtectedData = async ({
...
 }: IExecConsumer & ProcessProtectedDataParams): Promise<Taskid> => {
   try {
     const requester = await iexec.wallet.getAddress();
-    const vApp = addressOrEnsOrAnySchema()
+    const vApp = addressOrEnsSchema()
       .required()
       .label('authorizedApp')
       .validateSync(app);
-    const vProtectedData = addressOrEnsOrAnySchema()
+    const vProtectedData = addressOrEnsSchema()
```
Comment (PierreJeanjacquot) [inline packages/sdk/src/dataProtector/processProtectedData.ts:36]: this was a bug I found during the development, the protectedData must be specified in processProtectedData (cannot be "any")

File: packages/sdk/src/dataProtector/processProtectedData.ts [hunk 5]
```diff
@@ -24,14 +25,15 @@ export const processProtectedData = async ({
   args,
   inputFiles,
   secrets,
+  workerpool,
```
Comment (PierreJeanjacquot) [inline packages/sdk/src/dataProtector/processProtectedData.ts:28]: added `workerpool` option to override the default value as it was not possible (this is required for local testing as we don't control the default pool)

File: packages/sdk/src/dataProtector/protectDataObservable.ts [hunk 6]
```diff
@@ -140,7 +140,8 @@ export const protectDataObservable = ({
             vName,
             JSON.stringify(schema),
             multiaddrBytes,
-            checksum
+            checksum,
+            txOptions
```
Comment (PierreJeanjacquot) [inline packages/sdk/src/dataProtector/protectDataObservable.ts:144]: txOptions specifies the `gasPrice` we want to inherit it from the `iexec` SDK instance

File: packages/sdk/src/utils/validators.ts [hunk 7]
```diff
@@ -17,7 +17,8 @@ const isZeroStringTest = (value: string) => value === '0';
 export const stringSchema = () =>
   string().strict().typeError('${path} should be a string');
 
-export const urlSchema = () => string().url('${path} should be a url');
+export const urlSchema = () =>
+  string().matches(/^http[s]?:\/\//, '${path} should be a url');
```
Comment (PierreJeanjacquot) [inline packages/sdk/src/utils/validators.ts:21]: changed the url validation as `yup`'s one does not support `localhost`
the new implementation is not bulletproof but it provides a cheap quick validation

File: packages/sdk/tests/e2e/processProtectedData.test.ts [hunk 8]
```diff
@@ -1,29 +1,76 @@
...
+      });
+    await resourceProvider.order
+      .createWorkerpoolorder({
+        workerpool: workerpoolAddress,
+        category: 0,
+        tag: ['tee', 'scone'],
+      })
+      .then(resourceProvider.order.signWorkerpoolorder)
+      .then(resourceProvider.order.publishWorkerpoolorder);
+    app = appAddress;
```
Comment (PierreJeanjacquot) [inline packages/sdk/tests/e2e/processProtectedData.test.ts:66]: create resources on the local stack as the one used before required orders published by third parties on the marketplace (our marketplace instance is empty)

File: packages/sdk/tests/prepare-test-env.js
Comment (PierreJeanjacquot) [inline packages/sdk/tests/prepare-test-env.js:1]: fetch the latest block of the bellecour blockchain to base the local fork

File: packages/sdk/tests/test-utils.ts [hunk 9]
```diff
@@ -1,8 +1,46 @@
...
+    ipfsGateway: process.env.DRONE
+      ? 'http://ipfs:8080'
+      : 'http://localhost:8080',
+    ipfsNode: process.env.DRONE ? 'http://ipfs:5001' : 'http://localhost:5001',
+    subgraphUrl: process.env.DRONE
+      ? 'http://graphnode:8000/subgraphs/name/DataProtector'
+      : 'http://localhost:8000/subgraphs/name/DataProtector',
+  };
+  return [ethProvider, options];
+};
```
Comment (PierreJeanjacquot) [inline packages/sdk/tests/test-utils.ts:42]: create the configuration to connect the local test stack

File: packages/subgraph/deployer.Dockerfile
Comment (PierreJeanjacquot) [inline packages/subgraph/deployer.Dockerfile:1]: provide a base image for subgraph deployment in local stack, this allow us to cache dependencies installation across multiple reinstantiation of the local stack.

File: packages/subgraph/schema.graphql
Comment (PierreJeanjacquot) [inline packages/subgraph/schema.graphql:1]: some v2 code was left, rolled back to what it should be on develop branch (plus `Subscription` is reserved word that prevents to compile)

**Review by SeddikBellamine — CHANGES_REQUESTED** [review]
Great work !! , i left you some comments, I really like the idea to change the used stack to a local one and not use the real stack.

File: packages/sdk/.drone.yml [hunk 10]
```diff
@@ -7,6 +7,216 @@ trigger:
...
 steps:
+  ##### START LOCAL STACK #####
+  - name: prepare local-stack-env
+    image: node
+    commands:
+      - node packages/sdk/tests/prepare-test-env.js
+
+  - name: bellecour-fork
+    detach: true
+    image: ghcr.io/foundry-rs/foundry:latest
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml:18]: I initially started with a hardhat node but had some trouble when all the stack components were connected to it (node becoming unresponsive)
However, I haven't tested hardhat again after stabilizing with foundry anvil.
Another benefit I found after tinkering with foundry was the ability to pass all the fork parameters through cli options, freeing us of the additional burden of maintaining the hardhat docker image with bellecour fork configuration.

File: packages/sdk/.drone.yml [hunk 11]
```diff
@@ -7,6 +7,216 @@ trigger:
...
+      ipfs: ipfs:5001
+      ethereum: bellecour:http://bellecour-fork:8545
+    commands:
+      # bellecour-fork healthcheck
+      - while ! nc -z bellecour-fork 8545 ; do sleep 1 ; done && echo "bellecour-fork ready"
+      # ipfs healthcheck
+      - while ! nc -z ipfs 8080 ; do sleep 1 ; done && echo "ipfs gateway ready"
+      - while ! nc -z ipfs 5001 ; do sleep 1 ; done && echo "ipfs upload ready"
+      # graphnode-postgres healthcheck
+      - while ! nc -z graphnode-postgres 5432 ; do sleep 1 ; done && echo "graphnode-postgres ready"
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml:87]: I'm aware of `depends_on` and I would like to use it the way I did for the local docker-compose counterpart.
For what I tested drone does not support depends_on X healthy (only depends on X started) which is not enough in this case.
There is probably some room for optimization with depends_on, detached services, and blocking healthchecks steps, but using `depends_on` would convert the whole pipeline into a graph pipeline which is slightly more complex, I'll ask the team for approval :+1:

File: packages/sdk/.drone.yml [hunk 12]
```diff
@@ -7,6 +7,216 @@ trigger:
...
+      - docker-entrypoint.sh postgres -cshared_preload_libraries=pg_stat_statements
+    expose:
+      - 5432
+    environment:
+      POSTGRES_USER: graphnode
+      POSTGRES_PASSWORD: password
+      POSTGRES_DB: graphnode-db
+
+  - name: graphnode
+    image: graphprotocol/graph-node:v0.27.0
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml:67]: sure, I'll check what is used in production

File: packages/sdk/src/dataProtector/processProtectedData.ts [hunk 13]
```diff
@@ -88,7 +95,7 @@ export const processProtectedData = async ({
       appmaxprice: vMaxPrice,
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/src/dataProtector/processProtectedData.ts:95]: I think you found a bug :bug: :+1:

[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/src/dataProtector/processProtectedData.ts]: fixed 69a1eb2df073675ee245907fe9b03486c658b829

File: packages/sdk/src/utils/validators.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/src/utils/validators.ts:21]: We also need to support drone/docker service name here (ex: `http://graphnode:8000`)
The `/^http[s]?:\/\//` regex is not that big maybe not worth adding in a variable as it is not used elsewhere

File: packages/sdk/.drone.yml [hunk 14]
```diff
@@ -7,6 +7,216 @@ trigger:
...
+      - while ! nc -z market-mongo 27017 ; do sleep 1 ; done && echo "market-mongo ready"
+      # market-redis healthcheck
+      - while ! nc -z market-redis 6379 ; do sleep 1 ; done && echo "market-redis ready"
+      # start
+      - export START_BLOCK=$(cat local-stack-env/BELLECOUR_FORK_BLOCK)
+      - cd /app
+      - npm start
+
+  - name: local subgraph deployment
+    image: node:18
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml]: There is some unclear reason why `18.19` is used in other pipelines probably linked to a bug affecting `tsc`. Changing to 18.19 but only for the sake of having the same version. We need to challenge it.

changed 1aa2e3ad959ae4c8bac6f17e2cbf93cea84b587c

File: packages/sdk/.drone.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml:87]: @SeddikBellamine proposal for a graphpipeline https://github.com/iExecBlockchainComputing/dataprotector-sdk/pull/184

File: packages/sdk/.drone.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml:67]: checked with infra team, the deployed graphnodes uses `0.27.0`, nothing to change on our side IMO.

File: packages/sdk/.drone.yml (hunk 1 above)
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml]: changed in #184

File: packages/sdk/.drone.yml (hunk 2 above)
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml]: changed in #184

File: packages/sdk/.drone.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/.drone.yml]: changed in #184

**Review by SeddikBellamine — APPROVED** [review]
LGTM !

File: packages/sdk/.drone.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (SeddikBellamine) [reply packages/sdk/.drone.yml:112]: got it

**Pull request merged**